### PR TITLE
[Automation] RHEL-137166 - Run virt-who in FIPS mode

### PR DESCRIPTION
**Pull request by warku123**
**Description**
Add a automation test case to test ESXi available in FIPS mode.

**Test Result**
```
================================================================== test session starts ==================================================================
platform linux -- Python 3.9.13, pytest-6.2.5, py-1.11.0, pluggy-0.13.1
rootdir: /home/jzhang/文档/test_repos/virtwho-test, configfile: pytest.ini
plugins: anyio-3.5.0, services-1.3.1, mock-1.10.4, forked-1.6.0, xdist-1.34.0
collected 30 items / 29 deselected / 1 selected                                                                                                         

tests/hypervisor/test_esx.py .                                                                                                                    [100%]

=================================================================== warnings summary ====================================================================
../../../anaconda3/lib/python3.9/site-packages/paramiko/transport.py:216
  /home/jzhang/anaconda3/lib/python3.9/site-packages/paramiko/transport.py:216: CryptographyDeprecationWarning: Blowfish has been deprecated
    "class": algorithms.Blowfish,

tests/hypervisor/test_esx.py:1394
  /home/jzhang/文档/test_repos/virtwho-test/tests/hypervisor/test_esx.py:1394: PytestUnknownMarkWarning: Unknown pytest.mark.notStage - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.notStage

tests/hypervisor/test_esx.py:1438
  /home/jzhang/文档/test_repos/virtwho-test/tests/hypervisor/test_esx.py:1438: PytestUnknownMarkWarning: Unknown pytest.mark.notStage - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.notStage

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=============================================== 1 passed, 29 deselected, 3 warnings in 219.69s (0:03:39) ================================================
[2023-10-25 10:51:06] - [conftest.py] - INFO: Finished Test: tests/hypervisor/test_esx.py::TestEsxNegative::test_run_in_FIPS_mode
```

### CODE REVIEW

File: tests/hypervisor/test_esx.py [hunk 1]
```diff
@@ -1432,6 +1433,54 @@ def test_post_large_json_to_rhsm(self, register_data, ssh_host):
...
             logger.warning(output)
+    
+    @pytest.mark.tier2
+    @pytest.mark.notStage
+    def test_run_in_FIPS_mode(self, function_hypervisor, virtwho, ssh_host):
+        """
+        :title: virt-who: esx: test run virt-who in FIPS mode
+        :id: 2a05a366-6b92-456a-89ca-c7576491c0bb
+        :caseimportance: High
+        :tags: hypervisor,esx,tier2
```
Comment (hkx303) [inline tests/hypervisor/test_esx.py:1444]: should be tier3 I thinik

File: tests/hypervisor/test_esx.py [hunk 2]
```diff
@@ -1432,6 +1433,54 @@ def test_post_large_json_to_rhsm(self, register_data, ssh_host):
...
+        :expectedresults:
+            1. Succeeded to enable FIPS mode on the host
+        """
+        timeout = 300
+        interval = 5
+        
+        ssh_host.runcmd("fips-mode-setup --enable")
+        ssh_host.runcmd("sudo reboot")
+        start_time = time.time()
+        # 等待几秒，以便服务器开始重启过程
```
Comment (hkx303) [inline tests/hypervisor/test_esx.py:1462]: please translate it into English,  thanks

File: tests/hypervisor/test_esx.py [hunk 3]
```diff
@@ -1432,6 +1433,54 @@ def test_post_large_json_to_rhsm(self, register_data, ssh_host):
...
+            if elapsed_time > timeout:
+                assert False, f"Timeout reached after {timeout} seconds!"
+            print(f"Waiting for server to come back online... Elapsed time: {int(elapsed_time)} seconds.")
+            time.sleep(interval)
+        
+        _, stdout = ssh_host.runcmd("cat /proc/sys/crypto/fips_enabled")
+        assert("1" in stdout)
+        
+        result = virtwho.run_service()
+        print(result)
```
Comment (hkx303) [inline tests/hypervisor/test_esx.py:1475]: seems we don't need this line now

File: tests/hypervisor/test_esx.py [hunk 4]
```diff
@@ -1432,6 +1433,54 @@ def test_post_large_json_to_rhsm(self, register_data, ssh_host):
...
+        assert (
+            result["error"] == 0
+            and result["send"] == 1
+            and result["thread"] == 1
+        )
+        
+        ssh_host.runcmd("fips-mode-setup --disable")
+        ssh_host.runcmd("sudo reboot")
```
Comment (hkx303) [inline tests/hypervisor/test_esx.py:1483]: and we should make sure we change back by  "cat /proc/sys/crypto/fips_disabled"

File: tests/hypervisor/test_esx.py [hunk 5]
```diff
@@ -1432,6 +1433,54 @@ def test_post_large_json_to_rhsm(self, register_data, ssh_host):
...
+        
+        result = virtwho.run_service()
+        print(result)
+        assert (
+            result["error"] == 0
+            and result["send"] == 1
+            and result["thread"] == 1
+        )
+        
+        ssh_host.runcmd("fips-mode-setup --disable")
```
Comment (hkx303) [inline tests/hypervisor/test_esx.py:1482]: to confirm we change back to disable mode after this case, maybe we need to use the `try...finally..`. like case: https://github.com/VirtwhoQE/virtwho-test/blob/main/tests/package/test_upgrade_downgrade.py#L33
what do you think of this?

File: tests/hypervisor/test_esx.py [hunk 6]
```diff
@@ -1454,3 +1503,13 @@ def json_data_create(hypervisors_num, guests_num):
...
+
+def is_host_responsive(host):
+    """
+    Check if the host is responsive
+    :param host: host ip address
+    :return: True or False
+    """
+    print(host)
+    cmd = f"ping -c 1 {host}"
+    return os.system(cmd) == 0
```
Comment (hkx303) [inline tests/hypervisor/test_esx.py:1515]: add the new line at the end of file

**Review by hkx303 — COMMENTED** [review]
Nice job Pinge, a few comments, and please upload the case description and test result to the PR's description. Thanks

File: tests/hypervisor/test_esx.py (hunk 5 above)
Comment (warku123) [reply tests/hypervisor/test_esx.py:1481]: Maybe I think no need to use try...finally in this code. I don't know what to do in finally... sentences.

File: tests/hypervisor/test_esx.py [hunk 7]
```diff
@@ -1432,7 +1433,72 @@ def test_post_large_json_to_rhsm(self, register_data, ssh_host):
...
+        ssh_host.runcmd("fips-mode-setup --disable")
+        ssh_host.runcmd("sudo reboot")
+        start_time = time.time()
+        # Waiting for server to reboot
+        time.sleep(interval)
+        while not is_host_responsive(ssh_host.host):
+            elapsed_time = time.time() - start_time
+            if elapsed_time > timeout:
+                assert False, f"Timeout reached after {timeout} seconds!"
+            print(f"Waiting for server to come back online... Elapsed time: {int(elapsed_time)} seconds.")
```
Comment (hkx303) [inline tests/hypervisor/test_esx.py:1490]: use `logger.info()` instead of `print(）`

File: tests/hypervisor/test_esx.py (hunk 5 above)
Comment (hkx303) [reply tests/hypervisor/test_esx.py:1481]: we need to change back to FIPS disable mode, so we should add the `fips-mode-setup --disable` to finally

**Review by hkx303 — COMMENTED** [review]
Thanks for the quick reply, also a few comments... and don't forget to upload the test result and description to this PR

**Comment by warku123** [comment]
*Test Result*
```
================================================================== test session starts ==================================================================
platform linux -- Python 3.9.13, pytest-6.2.5, py-1.11.0, pluggy-0.13.1
rootdir: /home/jzhang/文档/test_repos/virtwho-test, configfile: pytest.ini
plugins: anyio-3.5.0, services-1.3.1, mock-1.10.4, forked-1.6.0, xdist-1.34.0
collected 30 items / 29 deselected / 1 selected                                                                                                         

tests/hypervisor/test_esx.py .                                                                                                                    [100%]

=================================================================== warnings summary ====================================================================
../../../anaconda3/lib/python3.9/site-packages/paramiko/transport.py:216
  /home/jzhang/anaconda3/lib/python3.9/site-packages/paramiko/transport.py:216: CryptographyDeprecationWarning: Blowfish has been deprecated
    "class": algorithms.Blowfish,

tests/hypervisor/test_esx.py:1394
  /home/jzhang/文档/test_repos/virtwho-test/tests/hypervisor/test_esx.py:1394: PytestUnknownMarkWarning: Unknown pytest.mark.notStage - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.notStage

tests/hypervisor/test_esx.py:1438
  /home/jzhang/文档/test_repos/virtwho-test/tests/hypervisor/test_esx.py:1438: PytestUnknownMarkWarning: Unknown pytest.mark.notStage - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.notStage

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=============================================== 1 passed, 29 deselected, 3 warnings in 219.69s (0:03:39) ================================================
[2023-10-25 10:51:06] - [conftest.py] - INFO: Finished Test: tests/hypervisor/test_esx.py::TestEsxNegative::test_run_in_FIPS_mode
```

**Comment by warku123** [comment]
@hkx303 I think it's all well now, I add the description in the first comment.

File: tests/hypervisor/test_esx.py [hunk 8]
```diff
@@ -1432,7 +1433,67 @@ def test_post_large_json_to_rhsm(self, register_data, ssh_host):
         else:
             logger.warning("Failed to post json to satellite")
             logger.warning(output)
-
+    
+    @pytest.mark.tier2
```
Comment (hkx303) [inline tests/hypervisor/test_esx.py:1437]: should be `@pytest.mark.tier3`
and remove the `@pytest.mark.notStage`, we should test this case both satellite and stage

**Review by hkx303 — APPROVED** [review]
Thanks pinge! Just one comment :)

**Comment by hkx303** [comment]
Thanks for the quick change

**Pull request merged**